src/modules/job-exec/checkpoint.c

```diff
@@ -0,0 +1,261 @@
+ * and the guest namespace is created and linked from the primary
+ * namespace. A guest.exec.eventlog is created with an initial "init"
+ * event posted.
+ *
+ * Jobspec and R are parsed as soon as asynchronous initialization tasks
+ * complete. If any of these steps fail, an exec initialization exception
+ * is thrown. Finally, the implementation "init" method is called.
+ *
+ * JOB STARTING/RUNNING:
+ *
+ * After initialization is complete, the exec service emits a "starting"
+ * event to the exec eventlog and calls the implementaton "start" method.
+ * Once all job shells or equivalent are running, the exec implementation
+ * should invoke jobinfo_started(), which emits a "running" event to the
+ * exec eventlog and sends the "start" response to the job-manager.
+ *
+ * JOB FINISH/CLEANUP:
+ *
+ * As tasks/job shells exit, the exec implementation should call
+ * jobinfo_tasks_complete(), which emits a "complete" event to the exec
+ * eventlog, sends a "finish" response to the job-manager.
+ *
+ * JOB FINALIZATION:
+ *
+ * jobinfo_finalize() is called after the "finish" event, which performs
+ * the following tasks:
+ *
+ *  - terminating "done" event is posted to the exec.eventlog
+ *  - the guest namespace, now quiesced, is copied to the primary namespace
+ *  - the guest namespace is removed
+ *  - the final "release final=true" response is sent to the job manager
+ *  - the local job object is destroyed
+ *
+ * TEST CONFIGURATION
+ *
+ * The job-exec module supports an object in the jobspec under
+ * attributes.system.exec.test, which enables mock execution and
+ * supports the following keys
+ *
+ * {
+ *   "run_duration":s,      - alternate/override attributes.system.duration
+ *   "wait_status":i        - report this value as status in the "finish" resp
+ *   "mock_exception":s     - mock an exception during this phase of job
+ *                             execution (currently "init" and "run")
+ * }
+ *
+ * The "bulk" execution implementation supports testing and other
+ * paramters under attributes.system.exec.bulkexec, including:
+ *
+ * {
+ *   "mock_exception":s     - cancel job after a certain number of shells
+ *                            have been launched.
+ * }
+ *
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <assert.h>
+#include <unistd.h>
+#include <flux/core.h>
+
+#include "src/common/libczmqcontainers/czmq_containers.h"
+
+#include "job-exec.h"
+#include "checkpoint.h"
+
+static flux_future_t *lookup_nsroots (flux_t *h, zhashx_t *jobs)
+{
+    struct jobinfo *job = zhashx_first (jobs);
+    flux_future_t *fall = NULL;
+    flux_future_t *f = NULL;
+
+    while (job) {
+        if (job->running) {
+            if (!fall) {
+                if (!(fall = flux_future_wait_all_create ())) {
+                    flux_log_error (h, "lookup_nsroots: "
+                                            "flux_future_wait_all_create");
+                    goto cleanup;
+                }
```

Comment:
Looks like all the error cases in this function are equally unlikely and uninteresting to distinguish.  Suggest you simply log one message down in `checkpoint_running()` where the function is called, like "failed to initiate KVS requests for namespace checkpoint".

Reply:
Was going to make the same comment for the function after that one.  Maybe whole file needs error logs audited.